File: code/ex139.java
// Backtracking + DP(memo)

import java.util.List;
import java.util.HashMap;

class Solution {
    public boolean backtracking(String s, List<String> wordDict, HashMap<String, Boolean> dp) {
        // Get value from dp dictionary
        if (dp.containsKey(s))    return dp.get(s);
        
        // Add to list if s itself is in the wordDict
        if (wordDict.contains(s))    return true;
        
        for (int i = 0; i < s.length(); i++) {
            String end = s.substring(i + 1, s.length());
            
            if (wordDict.contains(end)) {
                if (backtracking(s.substring(0, i + 1), wordDict, dp))    return true;
            }      
        }
        
        dp.put(s, false);    // Update dp dictionary
        return false;
    }
    
    public boolean wordBreak(String s, List<String> wordDict) {
        HashMap<String, Boolean> dp = new HashMap<>();
        
        return backtracking(s, wordDict, dp);
    }
}

File: code/ex139.cpp
// DP

class Solution {
public:
    bool wordBreak(string s, vector<string>& wordDict) {
        int n = s.length();
        
        unordered_set<string> set;
        for (string word : wordDict)    set.insert(word);
        
        // Init dp array
        vector<bool> dp(n + 1);
        dp[0] = true;
        
        // Build dp array
        for (int i = 0; i < n; i++) {
            if (!dp[i])    continue;
            for (int j = i; j < n; j++) {
                if (set.find(s.substr(i, j - i + 1)) != set.end())
                    dp[j + 1] = true;
            }
        }
        
        return dp[n];
    }
};